dworp/scheduling.py

```diff
@@ -60,3 +62,61 @@ def __init__(self, size, rng):
 
     def step(self, now, agents, env):
         return self.rng.permutation(len(agents))[:self.size]
+
+
+class PoissonScheduler(Scheduler):
+    """
+    Calculates update times for a set of users that will be chosen according to the model for Poisson arrivals.
+    Each user will wake up at exponentially distributed random wait times, with a uniform parameter lambda, that
+    is constant over time and constant across users. Times are all integer-valued (could be interpreted as seconds).
+
+    Args:
+        size (int): size of the sample (i.e., the number of actors/agents)
+        rng (numpy.random.RandomState): numpy random generator
+        t0: Overall period start time (we ensure nothing is scheduled for this time)
+        tN: Overall period stop time (exclusive)
+        lmda: Lambda parameter for the exponential samples
+    """
+    def __init__(self, size, rng, t0, tN, lmda):
+        self.size = size
+        self.rng = rng
+        self.t0 = t0
+        self.tN = tN
+        self.lmda = lmda
+        thisschedule = self.create_schedule(size, t0, tN, lmda)
+        self.schedule_dict = thisschedule
+
+    def step(self, now, agents, env):
+        try:
+            thistimeagentinds = self.schedule_dict[now]
+        except KeyError:
+            print("Error in PoissonScheduler, calling step with a time that is not in the schedule, %f" % (now))
+            raise("Error in PoissonScheduler, calling step with a time that is not in the schedule")
+        return thistimeagentinds
+
+    def create_schedule(self, size, t0, tN, lmda):
+        # note that the time_dict returned is an ordered dict with keys the integer-valued times
+        # Loop through all users and generate login times from their start time to tN
+        times = []
+        for i in range(0, size):
+            t = t0
+            counter = 0
+            while t < tN and counter < 1e4:
+                counter = counter + 1
+                new_increment = self.rng.exponential(scale=lmda)
+                integer_increment = round(new_increment)
+                if integer_increment > 0:
+                    t = t + integer_increment
+                    if t < tN:
+                        times.append((i, t))
```

Comment:
I'm not sure why we are calling this a Poisson process. Isn't this a Bernoulli process now? And wouldn't it be more efficient to implement it that way with a geometric distribution (for small lambda, this throws away 10-40% of the samples)?

Reply:
A Poisson process has exponential inter-arrival times. We are sampling inter-activity times with the exponential distribution, so that's why we are calling it Poisson.

Reply:
But these aren't exponential wait times. They are geometric wait times where 
![image](https://user-images.githubusercontent.com/199558/46634992-d784d980-cb20-11e8-9c51-0ab64198d45b.png)

Here's a sample plot at lambda=5 and p=e^-y (discretized exponential and geometric):

![image](https://user-images.githubusercontent.com/199558/46635238-a8bb3300-cb21-11e8-8f87-887d5fbd78bc.png)